{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "976b21a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml\n",
    "import pandas as pd\n",
    "from typing import Generator, Tuple, Dict\n",
    "import multiprocessing\n",
    "from typing import Optional"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8dd5d71f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta(filepath: str) -> Generator[Tuple[str, str], None, None]:\n",
    "    \"\"\"\n",
    "    Parses a FASTA file and yields record names and sequences.\n",
    "    Args:\n",
    "        filepath (str): Path to the FASTA file.\n",
    "    Yields:\n",
    "        tuple: A tuple containing the record name and sequence.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        with open(filepath, 'r') as file:\n",
    "            name, sequence = None, []\n",
    "            for line in file:\n",
    "                line = line.strip()\n",
    "                if line.startswith(\">\"):\n",
    "                    if name:\n",
    "                        yield name, ''.join(sequence)\n",
    "                    name, sequence = line[1:], []\n",
    "                else:\n",
    "                    sequence.append(line)\n",
    "            if name:\n",
    "                yield name, ''.join(sequence)\n",
    "    except FileNotFoundError:\n",
    "        raise FileNotFoundError(f\"File not found: {filepath}\")\n",
    "    except Exception as e:\n",
    "        raise RuntimeError(f\"An error occurred while reading the FASTA file: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2d820216",
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath = \"tests/test_fasta.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "50d37edf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGCGTACGTAGCTAGCGTAGCTAGT\n",
      "CGTAGCTAGTACGATCGTACGTAGCT\n"
     ]
    }
   ],
   "source": [
    "for _,seq in read_fasta(filepath):\n",
    "    print(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "c970a5ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def load_feature_data(prop: str, feature_file: str=\"data/lookup.yaml\") -> Dict[str, float]:\n",
    "    \"\"\"\n",
    "    Loads feature data from a YAML file for a given k-mer length.\n",
    "\n",
    "    Args:\n",
    "        feature_file (str): Path to the YAML file containing feature data.\n",
    "        kmer_len (int): Length of the k-mer.\n",
    "\n",
    "    Returns:\n",
    "        dict: A dictionary of feature data.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        with open(feature_file, 'r') as f:\n",
    "            data = yaml.safe_load(f)\n",
    "            if str(prop) in data:\n",
    "                return data[str(prop)]\n",
    "            else:\n",
    "                raise ValueError(f\"k-mer length {kmer_len} not found in feature data.\")\n",
    "    except FileNotFoundError:\n",
    "        raise FileNotFoundError(f\"Feature file not found: {feature_file}\")\n",
    "    except yaml.YAMLError as e:\n",
    "        raise RuntimeError(f\"An error occurred while parsing the YAML file: {e}\")\n",
    "    except Exception as e:\n",
    "        raise RuntimeError(f\"An error occurred while loading feature data: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "9c754a9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NPP': {'AAA': 36,\n",
       "  'AAC': 6,\n",
       "  'AAG': 6,\n",
       "  'AAT': 30,\n",
       "  'ACA': 6,\n",
       "  'ACC': 8,\n",
       "  'ACG': 8,\n",
       "  'ACT': 11,\n",
       "  'AGA': 9,\n",
       "  'AGC': 25,\n",
       "  'AGG': 8,\n",
       "  'AGT': 11,\n",
       "  'ATA': 13,\n",
       "  'ATC': 7,\n",
       "  'ATG': 18,\n",
       "  'ATT': 30,\n",
       "  'CAA': 9,\n",
       "  'CAC': 17,\n",
       "  'CAG': 2,\n",
       "  'CAT': 18,\n",
       "  'CCA': 8,\n",
       "  'CCC': 13,\n",
       "  'CCG': 2,\n",
       "  'CCT': 8,\n",
       "  'CGA': 31,\n",
       "  'CGC': 25,\n",
       "  'CGG': 2,\n",
       "  'CGT': 8,\n",
       "  'CTA': 18,\n",
       "  'CTC': 8,\n",
       "  'CTG': 2,\n",
       "  'CTT': 6,\n",
       "  'GAA': 12,\n",
       "  'GAC': 8,\n",
       "  'GAG': 8,\n",
       "  'GAT': 7,\n",
       "  'GCA': 13,\n",
       "  'GCC': 45,\n",
       "  'GCG': 25,\n",
       "  'GCT': 25,\n",
       "  'GGA': 5,\n",
       "  'GGC': 45,\n",
       "  'GGG': 13,\n",
       "  'GGT': 8,\n",
       "  'GTA': 6,\n",
       "  'GTC': 8,\n",
       "  'GTG': 17,\n",
       "  'GTT': 6,\n",
       "  'TAA': 20,\n",
       "  'TAC': 6,\n",
       "  'TAG': 18,\n",
       "  'TAT': 13,\n",
       "  'TCA': 8,\n",
       "  'TCC': 5,\n",
       "  'TCG': 31,\n",
       "  'TCT': 9,\n",
       "  'TGA': 8,\n",
       "  'TGC': 13,\n",
       "  'TGG': 8,\n",
       "  'TGT': 6,\n",
       "  'TTA': 20,\n",
       "  'TTC': 12,\n",
       "  'TTG': 9,\n",
       "  'TTT': 36},\n",
       " 'DNaseI': {'AAT': -0.28,\n",
       "  'ATT': -0.28,\n",
       "  'AAA': -0.274,\n",
       "  'TTT': -0.274,\n",
       "  'CCA': -0.246,\n",
       "  'TGG': -0.246,\n",
       "  'AAC': -0.205,\n",
       "  'GTT': -0.205,\n",
       "  'CCG': -0.136,\n",
       "  'CGG': -0.136,\n",
       "  'ATC': -0.11,\n",
       "  'GAT': -0.11,\n",
       "  'ACT': -0.183,\n",
       "  'AGT': -0.183,\n",
       "  'AAG': -0.081,\n",
       "  'CTT': -0.081,\n",
       "  'CGC': -0.077,\n",
       "  'GCG': -0.077,\n",
       "  'AGG': -0.057,\n",
       "  'CCT': -0.057,\n",
       "  'GAA': -0.037,\n",
       "  'TTC': -0.037,\n",
       "  'ACG': -0.033,\n",
       "  'CGT': -0.033,\n",
       "  'ACC': -0.032,\n",
       "  'GGT': -0.032,\n",
       "  'GAC': -0.013,\n",
       "  'GTC': -0.013,\n",
       "  'CCC': -0.012,\n",
       "  'GGG': -0.012,\n",
       "  'ACA': -0.006,\n",
       "  'TGT': -0.006,\n",
       "  'CGA': -0.003,\n",
       "  'TCG': -0.003,\n",
       "  'GGA': 0.013,\n",
       "  'TCC': 0.013,\n",
       "  'CAA': 0.015,\n",
       "  'TTG': 0.015,\n",
       "  'AGC': 0.017,\n",
       "  'GCT': 0.017,\n",
       "  'GTA': 0.025,\n",
       "  'TAC': 0.025,\n",
       "  'AGA': 0.027,\n",
       "  'TCT': 0.027,\n",
       "  'CTC': 0.031,\n",
       "  'GAG': 0.031,\n",
       "  'CAC': 0.04,\n",
       "  'GTG': 0.04,\n",
       "  'TAA': 0.068,\n",
       "  'TTA': 0.068,\n",
       "  'GCA': 0.076,\n",
       "  'TGC': 0.076,\n",
       "  'CTA': 0.09,\n",
       "  'TAG': 0.09,\n",
       "  'GCC': 0.107,\n",
       "  'GGC': 0.107,\n",
       "  'ATG': 0.134,\n",
       "  'CAT': 0.134,\n",
       "  'CAG': 0.175,\n",
       "  'CTG': 0.175,\n",
       "  'ATA': 0.182,\n",
       "  'TAT': 0.182,\n",
       "  'TCA': 0.194,\n",
       "  'TGA': 0.194},\n",
       " 'bendabilityDNase': {'AAA': 0.1,\n",
       "  'AAC': 1.6,\n",
       "  'AAG': 4.2,\n",
       "  'AAT': 0,\n",
       "  'ACA': 5.8,\n",
       "  'ACC': 5.2,\n",
       "  'ACG': 5.2,\n",
       "  'ACT': 2,\n",
       "  'AGA': 6.5,\n",
       "  'AGC': 6.3,\n",
       "  'AGG': 4.7,\n",
       "  'AGT': 2,\n",
       "  'ATA': 9.7,\n",
       "  'ATC': 3.6,\n",
       "  'ATG': 8.7,\n",
       "  'ATT': 0,\n",
       "  'CAA': 6.2,\n",
       "  'CAC': 6.8,\n",
       "  'CAG': 9.6,\n",
       "  'CAT': 8.7,\n",
       "  'CCA': 0.7,\n",
       "  'CCC': 5.7,\n",
       "  'CCG': 3,\n",
       "  'CCT': 4.7,\n",
       "  'CGA': 5.8,\n",
       "  'CGC': 4.3,\n",
       "  'CGG': 3,\n",
       "  'CGT': 5.2,\n",
       "  'CTA': 7.8,\n",
       "  'CTC': 6.6,\n",
       "  'CTG': 9.6,\n",
       "  'CTT': 4.2,\n",
       "  'GAA': 5.1,\n",
       "  'GAC': 5.6,\n",
       "  'GAG': 6.6,\n",
       "  'GAT': 3.6,\n",
       "  'GCA': 7.5,\n",
       "  'GCC': 8.2,\n",
       "  'GCG': 4.3,\n",
       "  'GCT': 6.3,\n",
       "  'GGA': 6.2,\n",
       "  'GGC': 8.2,\n",
       "  'GGG': 5.7,\n",
       "  'GGT': 5.2,\n",
       "  'GTA': 6.4,\n",
       "  'GTC': 5.6,\n",
       "  'GTG': 6.8,\n",
       "  'GTT': 1.6,\n",
       "  'TAA': 7.3,\n",
       "  'TAC': 6.4,\n",
       "  'TAG': 7.8,\n",
       "  'TAT': 9.7,\n",
       "  'TCA': 10,\n",
       "  'TCC': 6.2,\n",
       "  'TCG': 5.8,\n",
       "  'TCT': 6.5,\n",
       "  'TGA': 10,\n",
       "  'TGC': 7.5,\n",
       "  'TGG': 0.7,\n",
       "  'TGT': 5.8,\n",
       "  'TTA': 7.3,\n",
       "  'TTC': 5.1,\n",
       "  'TTG': 6.2,\n",
       "  'TTT': 0.1},\n",
       " 'bendabilityConcensus': {'AAA': 0.05,\n",
       "  'AAC': 2.65,\n",
       "  'AAG': 4.7,\n",
       "  'AAT': 0.35,\n",
       "  'ACA': 5.5,\n",
       "  'ACC': 5.3,\n",
       "  'ACG': 5.3,\n",
       "  'ACT': 3.9,\n",
       "  'AGA': 4.9,\n",
       "  'AGC': 6.9,\n",
       "  'AGG': 5.05,\n",
       "  'AGT': 3.9,\n",
       "  'ATA': 6.25,\n",
       "  'ATC': 4.45,\n",
       "  'ATG': 7.7,\n",
       "  'ATT': 0.35,\n",
       "  'CAA': 4.75,\n",
       "  'CAC': 6.65,\n",
       "  'CAG': 6.9,\n",
       "  'CAT': 7.7,\n",
       "  'CCA': 3.05,\n",
       "  'CCC': 5.85,\n",
       "  'CCG': 3.85,\n",
       "  'CCT': 5.05,\n",
       "  'CGA': 7.05,\n",
       "  'CGC': 5.9,\n",
       "  'CGG': 3.85,\n",
       "  'CGT': 5.3,\n",
       "  'CTA': 5,\n",
       "  'CTC': 6,\n",
       "  'CTG': 6.9,\n",
       "  'CTT': 4.7,\n",
       "  'GAA': 4.05,\n",
       "  'GAC': 5.5,\n",
       "  'GAG': 6,\n",
       "  'GAT': 4.45,\n",
       "  'GCA': 6.75,\n",
       "  'GCC': 9.1,\n",
       "  'GCG': 5.9,\n",
       "  'GCT': 6.9,\n",
       "  'GGA': 5,\n",
       "  'GGC': 9.1,\n",
       "  'GGG': 5.85,\n",
       "  'GGT': 5.3,\n",
       "  'GTA': 5.05,\n",
       "  'GTC': 5.5,\n",
       "  'GTG': 6.65,\n",
       "  'GTT': 2.65,\n",
       "  'TAA': 4.65,\n",
       "  'TAC': 5.05,\n",
       "  'TAG': 5,\n",
       "  'TAT': 6.25,\n",
       "  'TCA': 7.7,\n",
       "  'TCC': 5,\n",
       "  'TCG': 7.05,\n",
       "  'TCT': 4.9,\n",
       "  'TGA': 7.7,\n",
       "  'TGC': 6.75,\n",
       "  'TGG': 3.05,\n",
       "  'TGT': 5.5,\n",
       "  'TTA': 4.65,\n",
       "  'TTC': 4.05,\n",
       "  'TTG': 4.75,\n",
       "  'TTT': 0.05}}"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_lookup = load_feature_data(prop = \"trinucleotide\", \n",
    "                  feature_file=\"DNAflexpy/data/lookup.yaml\")\n",
    "feature_lookup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "20587239",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transform_seq_to_feat(sequence, kmer_len, feature, feature_lookup):\n",
    "    \"\"\"\n",
    "    Calculates average structural profile values for a given sequence window.\n",
    "\n",
    "    Args:\n",
    "        sequence (str): DNA sequence to process.\n",
    "        kmer_len (int): Length of k-mers.\n",
    "        feature (str): Feature to calculate.\n",
    "        feature_lookup (dict): Dictionary containing feature data.\n",
    "\n",
    "    Returns:\n",
    "        float: Average feature value for the given sequence window.\n",
    "    \"\"\"\n",
    "    sequence = sequence.upper()\n",
    "    feature_data = feature_lookup.get(feature, {})\n",
    "\n",
    "    if not feature_data:\n",
    "        print(f\"**FATAL** ==>{feature}<== not in lookup data (.yaml file). Check spelling or add it.\")\n",
    "        return 0\n",
    "\n",
    "    ls_values_w = []\n",
    "    for i in range(len(sequence) - kmer_len + 1):\n",
    "        subseq = sequence[i: i + kmer_len]\n",
    "        value = feature_data.get(subseq)\n",
    "        if value is not None:\n",
    "            ls_values_w.append(value)\n",
    "        else:\n",
    "            print(f\"**Warning** Subsequence {subseq} not found in the dictionary for {feature}\")\n",
    "\n",
    "    avg_w = sum(ls_values_w) / (len(sequence) - 1) if ls_values_w else 0\n",
    "    return avg_w\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c22474e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-0.026800000000000004"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transform_seq_to_feat(sequence=\"AATGCCATTAGATTAG\", kmer_len=3, feature=\"DNaseI\", feature_lookup=feature_lookup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "79412005",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq_to_numeric_profile(sequence, \n",
    "                        window_size, \n",
    "                        kmer_len, \n",
    "                        feature, \n",
    "                        feature_lookup):\n",
    "    \"\"\"\n",
    "        Desc:\n",
    "            - Operates on a sequence to aggregate average feature value of all the overlapping window\n",
    "\n",
    "        arguments:\n",
    "            - sequence\n",
    "            - window size to make overlapping window\n",
    "            - k-mer length\n",
    "            - feature: Feature to calculate\n",
    "            - feature_lookup: Dictionary containing feature data\n",
    "            \n",
    "        returns:\n",
    "            - a list of values for a sequence\n",
    "\n",
    "    \"\"\"\n",
    "    ls_window_avg_for_seq = []\n",
    "    for w_start in range(len(sequence) - window_size + 1):\n",
    "        current_w_seq = sequence[w_start : w_start + window_size]\n",
    "        avg_w = transform_seq_to_feat(current_w_seq, kmer_len, feature, feature_lookup)\n",
    "        ls_window_avg_for_seq.append(round(avg_w, 3))\n",
    "    \n",
    "    return ls_window_avg_for_seq\n",
    "\n",
    "def process_sequence(record, \n",
    "                    window_size, \n",
    "                    prop, \n",
    "                    feature, \n",
    "                    feature_lookup):\n",
    "    \"\"\"\n",
    "        Takes a sequence, \n",
    "        window size,\n",
    "        kmer length,\n",
    "        feature,\n",
    "        feature_lookup\n",
    "        \n",
    "    \"\"\"\n",
    "    current_seq = record.seq\n",
    "    feature_value = seq_to_numeric_profile(current_seq, window_size, kmer_len, feature, feature_lookup)\n",
    "    return feature_value\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "47bbb32c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[-0.018,\n",
       " 0.079,\n",
       " -0.016,\n",
       " -0.001,\n",
       " -0.098,\n",
       " -0.02,\n",
       " -0.031,\n",
       " 0.046,\n",
       " 0.002,\n",
       " -0.091,\n",
       " -0.081,\n",
       " -0.031]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_to_numeric_profile(sequence=\"AATGCCATTAGATTAG\", \n",
    "                        window_size=5, \n",
    "                        kmer_len=3, \n",
    "                        feature='DNaseI', \n",
    "                        feature_lookup=feature_lookup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "924f730c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_sequence(record, \n",
    "                    window_size, \n",
    "                    kmer_len, \n",
    "                    feature, \n",
    "                    feature_lookup):\n",
    "    \"\"\"\n",
    "        Takes a sequence, \n",
    "        window size,\n",
    "        kmer length,\n",
    "        feature,\n",
    "        feature_lookup\n",
    "        \n",
    "    \"\"\"\n",
    "    feature_value = seq_to_numeric_profile(record, window_size, kmer_len, feature, feature_lookup)\n",
    "    return feature_value\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "e2c9bb6c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.033, -0.009, -0.021, 0.004, 0.004, -0.01, -0.01, 0.02, 0.033, 0.031, 0.031, 0.049, 0.049, 0.007, -0.023, -0.021, 0.02, 0.033, 0.031, 0.031, 0.049, -0.001]\n",
      "[0.02, 0.033, 0.031, 0.031, 0.049, -0.001, -0.017, -0.033, 0.004, -0.003, -0.036, -0.056, -0.056, -0.036, -0.003, 0.004, 0.004, -0.01, -0.01, 0.02, 0.033, 0.031]\n"
     ]
    }
   ],
   "source": [
    "for _,seq in read_fasta(filepath):\n",
    "    t = process_sequence(record = seq, window_size=5, kmer_len=3, feature='DNaseI', feature_lookup= feature_lookup)\n",
    "    print(t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "46d0bb98",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_bendability(input_file: str, window_size: int, prop: str, kmer_len: int, feature: str, feature_file: str, threads: int, outfile: Optional[str] = None) -> Optional[pd.DataFrame]:\n",
    "    \"\"\"\n",
    "    Main function to calculate feature profiles from a FASTA file.\n",
    "\n",
    "    Args:\n",
    "        input_file (str): Path to the input multifasta file.\n",
    "        window_size (int): Size of the processing window.\n",
    "        prop (int): Dinucleotide/Trinucleotide.\n",
    "        feature (str): Feature to calculate.\n",
    "        feature_file (str): Path to the YAML file containing feature data.\n",
    "        threads (int): Number of threads to use.\n",
    "        outfile (str, optional): Output file name.\n",
    "\n",
    "    Returns:\n",
    "        Optional[pd.DataFrame]: Feature profiles if outfile is not specified, otherwise None.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Load feature data\n",
    "        feature_lookup = load_feature_data(prop, feature_file)\n",
    "        \n",
    "        # Parse the fasta file and process sequences\n",
    "#         records = list(read_fasta(input_file))\n",
    "        with multiprocessing.Pool(processes=threads) as pool:\n",
    "            results = pool.starmap(process_sequence, [(seq, window_size, kmer_len, feature, feature_lookup) for _, seq in read_fasta(input_file)])\n",
    "\n",
    "        # Create DataFrame from results\n",
    "        df = pd.DataFrame(results)\n",
    "        \n",
    "        if outfile:\n",
    "            df.to_csv(outfile, index=False, header=False, sep=\"\\t\")\n",
    "            logging.info(\"Process completed! Output written to %s\", outfile)\n",
    "        else:\n",
    "            return df  # Return DataFrame if no output file is specified\n",
    "    except Exception as e:\n",
    "        logging.error(\"An error occurred: %s\", e)\n",
    "        raise e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "ef668214",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Process SpawnPoolWorker-19:\n",
      "Process SpawnPoolWorker-20:\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/upalabdha/mambaforge/envs/test_DNAflexpy/lib/python3.13/multiprocessing/process.py\", line 313, in _bootstrap\n",
      "    self.run()\n",
      "    ~~~~~~~~^^\n",
      "  File \"/Users/upalabdha/mambaforge/envs/test_DNAflexpy/lib/python3.13/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/upalabdha/mambaforge/envs/test_DNAflexpy/lib/python3.13/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "  File \"/Users/upalabdha/mambaforge/envs/test_DNAflexpy/lib/python3.13/multiprocessing/queues.py\", line 387, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ~~~~~~~~~~~~~~~~~~~~~^^^^^\n",
      "AttributeError: Can't get attribute 'process_sequence' on <module '__main__' (<class '_frozen_importlib.BuiltinImporter'>)>\n",
      "  File \"/Users/upalabdha/mambaforge/envs/test_DNAflexpy/lib/python3.13/multiprocessing/process.py\", line 313, in _bootstrap\n",
      "    self.run()\n",
      "    ~~~~~~~~^^\n",
      "  File \"/Users/upalabdha/mambaforge/envs/test_DNAflexpy/lib/python3.13/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/upalabdha/mambaforge/envs/test_DNAflexpy/lib/python3.13/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "  File \"/Users/upalabdha/mambaforge/envs/test_DNAflexpy/lib/python3.13/multiprocessing/queues.py\", line 387, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "           ~~~~~~~~~~~~~~~~~~~~~^^^^^\n",
      "AttributeError: Can't get attribute 'process_sequence' on <module '__main__' (<class '_frozen_importlib.BuiltinImporter'>)>\n",
      "Process SpawnPoolWorker-22:\n",
      "Process SpawnPoolWorker-21:\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[86], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mlogging\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m test \u001b[38;5;241m=\u001b[39m \u001b[43mcalculate_bendability\u001b[49m\u001b[43m(\u001b[49m\u001b[43minput_file\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtests/test_fasta.fa\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[43m                   \u001b[49m\u001b[43mwindow_size\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m5\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[43m                   \u001b[49m\u001b[43mprop\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtrinucleotide\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      5\u001b[0m \u001b[43m                   \u001b[49m\u001b[43mkmer_len\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m3\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[43m                   \u001b[49m\u001b[43mfeature\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mDNaseI\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[43m                   \u001b[49m\u001b[43mfeature_file\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mDNAflexpy/data/lookup.yaml\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[43m                   \u001b[49m\u001b[43mthreads\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[82], line 24\u001b[0m, in \u001b[0;36mcalculate_bendability\u001b[0;34m(input_file, window_size, prop, kmer_len, feature, feature_file, threads, outfile)\u001b[0m\n\u001b[1;32m     21\u001b[0m         \u001b[38;5;66;03m# Parse the fasta file and process sequences\u001b[39;00m\n\u001b[1;32m     22\u001b[0m \u001b[38;5;66;03m#         records = list(read_fasta(input_file))\u001b[39;00m\n\u001b[1;32m     23\u001b[0m         \u001b[38;5;28;01mwith\u001b[39;00m multiprocessing\u001b[38;5;241m.\u001b[39mPool(processes\u001b[38;5;241m=\u001b[39mthreads) \u001b[38;5;28;01mas\u001b[39;00m pool:\n\u001b[0;32m---> 24\u001b[0m             results \u001b[38;5;241m=\u001b[39m \u001b[43mpool\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstarmap\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprocess_sequence\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[43m(\u001b[49m\u001b[43mseq\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mwindow_size\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkmer_len\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfeature\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfeature_lookup\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m_\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mseq\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mread_fasta\u001b[49m\u001b[43m(\u001b[49m\u001b[43minput_file\u001b[49m\u001b[43m)\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     26\u001b[0m         \u001b[38;5;66;03m# Create DataFrame from results\u001b[39;00m\n\u001b[1;32m     27\u001b[0m         df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(results)\n",
      "File \u001b[0;32m~/mambaforge/envs/test_DNAflexpy/lib/python3.13/multiprocessing/pool.py:375\u001b[0m, in \u001b[0;36mPool.starmap\u001b[0;34m(self, func, iterable, chunksize)\u001b[0m\n\u001b[1;32m    369\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mstarmap\u001b[39m(\u001b[38;5;28mself\u001b[39m, func, iterable, chunksize\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    370\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m'''\u001b[39;00m\n\u001b[1;32m    371\u001b[0m \u001b[38;5;124;03m    Like `map()` method but the elements of the `iterable` are expected to\u001b[39;00m\n\u001b[1;32m    372\u001b[0m \u001b[38;5;124;03m    be iterables as well and will be unpacked as arguments. Hence\u001b[39;00m\n\u001b[1;32m    373\u001b[0m \u001b[38;5;124;03m    `func` and (a, b) becomes func(a, b).\u001b[39;00m\n\u001b[1;32m    374\u001b[0m \u001b[38;5;124;03m    '''\u001b[39;00m\n\u001b[0;32m--> 375\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_map_async\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43miterable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstarmapstar\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mchunksize\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/mambaforge/envs/test_DNAflexpy/lib/python3.13/multiprocessing/pool.py:768\u001b[0m, in \u001b[0;36mApplyResult.get\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    767\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(\u001b[38;5;28mself\u001b[39m, timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[0;32m--> 768\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    769\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mready():\n\u001b[1;32m    770\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTimeoutError\u001b[39;00m\n",
      "File \u001b[0;32m~/mambaforge/envs/test_DNAflexpy/lib/python3.13/multiprocessing/pool.py:765\u001b[0m, in \u001b[0;36mApplyResult.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    764\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwait\u001b[39m(\u001b[38;5;28mself\u001b[39m, timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[0;32m--> 765\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_event\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/mambaforge/envs/test_DNAflexpy/lib/python3.13/threading.py:659\u001b[0m, in \u001b[0;36mEvent.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    657\u001b[0m signaled \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_flag\n\u001b[1;32m    658\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m signaled:\n\u001b[0;32m--> 659\u001b[0m     signaled \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_cond\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    660\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m signaled\n",
      "File \u001b[0;32m~/mambaforge/envs/test_DNAflexpy/lib/python3.13/threading.py:359\u001b[0m, in \u001b[0;36mCondition.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    357\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:    \u001b[38;5;66;03m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[39;00m\n\u001b[1;32m    358\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 359\u001b[0m         \u001b[43mwaiter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43macquire\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    360\u001b[0m         gotit \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    361\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import logging\n",
    "test = calculate_bendability(input_file= \"tests/test_fasta.fa\", \n",
    "                   window_size = 5,\n",
    "                   prop = \"trinucleotide\",\n",
    "                   kmer_len= 3, \n",
    "                   feature = 'DNaseI',\n",
    "                   feature_file = 'DNAflexpy/data/lookup.yaml',\n",
    "                   threads = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29d65775",
   "metadata": {},
   "outputs": [],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcbe8df3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
